tests/remotes.py

```diff
@@ -0,0 +1,337 @@
+import os
```

Comment:
Several high level considerations here:

1. This is cool, this was waiting to be factored out for quite some time.
2. Are you sure you want to do that in this PR? You may come across some issues while unifying this that you don't expect now. 
3. Things like `_should_test_aws()` and `get_ssh_url()` should be merged into classes, e.g. move to `S3.should_test()` and `SSH.get_url()`.

Reply:
> Are you sure you want to do that in this PR? You may come across some issues while unifying this that you don't expect now.

This PR has gotten quite large. So, I'd say, I can work on *nitpicky* changes on a separate PR. 
If it's a blocker, I'd definitely fix those, though. The tests have all passed. So, I guess, there won't be further issues.

> Things like _should_test_aws() and get_ssh_url() should be merged into classes, e.g. move to S3.should_test() and SSH.get_url().
Can I do that refactor in a separate PR? I want just wanted to move things around for this PR, so, I'd opt not to merge those in this PR itself.

P.S. I haven't also added tests as it will collide with #2873, which, I'll do in a separate PR.

Reply:
You may postpone some things, just make an issue for it and refer to a comment/discussion here.

Reply:
Can you review again?  :)

Reply:
> Things like _should_test_aws() and get_ssh_url() should be merged into classes, e.g. move to S3.should_test() and SSH.get_url().

Created #2878 to work on this further.